packages/contract/lib/promievent.js

```diff
@@ -1,36 +1,107 @@
+const debug = require("debug")("contract:promievent");
+const DebugUtils = require("@truffle/debug-utils");
 const Web3PromiEvent = require("web3-core-promievent");
 
-function PromiEvent() {
-  const { resolve, reject, eventEmitter } = Web3PromiEvent.apply(
-    null,
-    arguments
-  );
+function PromiEvent(justPromise, bugger) {
+  //selectors are passed in separately because when I
+  //try to import Debugger here I get a compile error
+  const { resolve, reject, eventEmitter } = new Web3PromiEvent(justPromise);
 
-  const originalStackTrace = new Error().stack;
+  originalStackTrace = new Error().stack;
 
   function rejectHijacker(e) {
-    try {
-      const stackTrace = originalStackTrace.replace(
-        /^Error: \n/,
-        e.stack.split("\n")[0]
-      );
-
-      e.hijackedStack = e.stack;
-      e.stack = stackTrace;
-    } catch (_) {
-      // Ignore the case when the replacement fails
+    debug("hijacking!");
+    debug("hash: %s", this.txHash);
+    let getSolidityStackTrace;
+    if (bugger && this.txHash) {
+      debug("debugging time!");
+      getSolidityStackTrace = async () => {
+        try {
+          let session = bugger.connect();
+          await session.load(this.txHash);
+          await session.continueUntilBreakpoint();
+          const report = session.view(
+            bugger.constructor.selectors.stacktrace.current.finalReport
+          );
+          const rawMessage = session.view(
+            bugger.constructor.selectors.evm.current.step.returnValue
+          );
+          await session.unload();
```

Comment:
Have you considered making this a new Session method? Seems like it might make sense, to avoid exposing all these details.

Reply:
(or maybe a Debugger method instead... `bugger.getStackTrace()` or something better named)

Reply:
This is a good idea, I think.  Could also add `getRevertMessage`.  Then I could simplify some of the logic currently sitting outside the debugger used for handling this stuff.

Reply:
These would be session methods, not debugger methods, because that's where we put these things.  That said, I do wonder why we have the separate debugger/session mechanism at all.  Like, why doesn't `forTx` (or `forProject`) just return a session in the first place so you don't have to do `bugger.connect()`?  It just seems like a pointless extra step.  (I'd say this is a breaking change, but I guess it could be made non-breaking if we gave session a `connect` method that just returned `this`.)

Reply:
> It just seems like a pointless extra step.

I concur. Feel free to make this change!

Reply:
Note: I think I'm actually not going to add `getRevertMessage` for now for technical reasons.  Rather I'm going to add the revert message to the stacktrace object.  Longer explanation later.